R/S3.R

```diff
@@ -124,38 +124,94 @@ is_S3_class <- function(x) {
 }
 
 # -------------------------------------------------------------------------
-# Define a few base examples
+# Pull out validation functions so hit by code coverage
 
-S3_factor <- new_S3_class("factor",
-  function(.data = integer(), levels = character()) {
+validate_factor <- function(self) {
+  c(
+    if (typeof(self) != "integer")
+      "Underlying data must be an <integer>",
+    if (!is.character(attr(self, "levels")))
+      "attr(, 'levels') must be a <character>"
+  )
+}
+
+validate_date <- function(self) {
+  if (!is.numeric(self)) {
+    "Underlying data must be numeric"
+  }
+}
+
+validate_POSIXct <- function(self) {
+  if (!is.numeric(self)) {
+    return("Underlying data must be numeric")
+  }
+
+  tz <- attr(self, "tz")
+  if (!is.character(tz) || length(tz) != 1) {
+    return("attr(, 'tz') must be a single string")
+  }
+}
+
+validate_data.frame <- function(self) {
+  if (!is.list(self)) {
+    return("Underlying data must be a <list>")
+  }
+
+  rn <- attr(self, "row.names")
+  if (!is.character(rn) && !is.integer(rn)) {
+    return("attr(, 'row.names') must be a <character> or <integer>")
+  }
+
+  if (length(self) >= 1) {
+    ns <- unique(c(lengths(self), length(rn)))
```

Comment:
The above point likely affects this line because `rn` might be an integer vector of length 2 in the compact form